### ISSUE TITLE
What user or permissions for it when running the worker?

### ISSUE THREAD

**Issue by christiaanwesterbeek** [graphile/worker#151]
### Summary

I made a worker pretty much straight from your [quick-start example](https://github.com/graphile/worker#quickstart-library). I use schema `later`. It all works fine when I use a connection string with the master user of my RDS Postgres instance. I would like the run the worker with another user instead of the master user. When I do that, on a fresh database without the `later` schema yet, I get

> permission denied for database

By analysing the [sql statements](https://github.com/graphile/worker/tree/main/sql) that `run` executes to create the schema, I worked out all the permissions that are required. 

```javascript
create role ${DB_WORKER_USERNAME} login password '${DB_WORKER_PASSWORD}';

grant connect on database ${PGDATABASE} to ${DB_WORKER_USERNAME};
grant usage on schema later to ${DB_WORKER_USERNAME};

grant select, insert, update, delete on all tables in schema later to ${DB_WORKER_USERNAME};
grant execute on all functions in schema later to ${DB_WORKER_USERNAME};

-- for future tables that graphile/worker might add
alter default privileges in schema later
    grant select, insert, update, delete on tables to ${DB_WORKER_USERNAME};
alter default privileges in schema later
    grant execute on functions to ${DB_WORKER_USERNAME};
```

Took me a while to arrive here, but all the while I wasn't really convinced that this is the way to go. Especially since no one else in the issues mentioned "permission denied".

Then I thought, let's do the first run of the quick start example with the master user and then change it to my worker user. But that didn't work out because of RLS on the jobs table and it doesn't get any job records. But I could probably fix that by adding a policy for that worker user.

**TL;DR**
I am not sure what user I should use to run the workers. I now use the master user of my Postgres RDS instance, but i guess I shouldn't because... it's master. Trying another user took me a while and it was a bit painful and not finished yet. Is the above the complete set of permissions? Is either of these the correct approach? Or maybe I should run the worker as master but then switch roles in the queries that the tasks execute (like how postgraphile and postgrest switch from the authenticator user to another). Or another approach all together. Feel a bit stupid asking this ~since nobody mentioned~ or maybe I am trying too hard to avoid the master user?

### Additional context

Node.js v14, AWS RDS Postres v11, graphile-worker 0.8.1 and also Postgrest.

Found a related comment in issue #25

> > It is still weird to me, that the database role that can insert jobs is the one, that firstly created the schema on the first start.
> 
> I've not come up with a clean solution for using different privileges; in future I think different tasks may use different roles/settings, but we've not advanced the project that far yet. If you come up with concrete proposals for this, I'd love to hear them 👍

**Comment by benjie** [comment]
Graphile Worker is intended to be ran as the "database owner" role (it doesn't need to be superuser). Jobs are expected to execute in a privileged way, and if they need to change into a lower privilege it's the task's responsibility to accomplish that. Because there's no way for Graphile Worker to know whether a particular job is allowed to be queued by a particular application user (it has no concept of application users, keep in mind Graphile Worker and PostGraphile are completely independent projects that happen to pair nicely), if we didn't require elevated privileges it would be possible for user A to read or create jobs on behalf of user B which could cause all sorts of security issues. When you need your users to create jobs, they should do so via `SECURITY DEFINER` functions that you define which delegate the task to Graphile Worker after doing whatever checks are necessary. Think of it a bit like `sudo`.

If you fancied writing some of this up into the README to stop future people tripping over it, that would be appreciated :raised_hands: 

[semi-automated message] Thanks for your question; hopefully we're well on the way to helping you solve your issue. This doesn't currently seem to be a bug in the library so I'm going to close the issue, but please feel free to keep requesting help below and if it does turn out to be a bug we can definitely re-open it 👍

You can also ask for help in the #help-and-support channel in [our Discord chat](http://discord.gg/graphile).
